{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GPcounts with zero-inflated negative binomial likelihood"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nuha BinTayyash, 2020\n",
    "\n",
    "This notebook comapre GPcount fit with zero-inflated negative binomial, negative binomial and Gaussian likelihoods on ScRNA-seq gene expression data for highly expressed genes in Islet $\\alpha$ cell from [GSE8737 single cell RNA-seq ](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87375) dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gpflow "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load $\\alpha$ dataset and pseudotime information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Y = pd.read_csv('normalized_alpha_counts.csv',index_col=[0])\n",
    "X = pd.read_csv('alpha_time_points.csv',index_col=[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fit GPcount with zero inflated negative binomial likelihood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from GPcounts.GPcounts_Module import Fit_GPcounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = ['ENSMUSG00000038943', 'ENSMUSG00000000394', 'ENSMUSG00000039396',\n",
    "       'ENSMUSG00000040856', 'ENSMUSG00000023944', 'ENSMUSG00000062510',\n",
    "       'ENSMUSG00000012443', 'ENSMUSG00000066366', 'ENSMUSG00000017716',\n",
    "       'ENSMUSG00000027326', 'ENSMUSG00000027447', 'ENSMUSG00000021270',\n",
    "       'ENSMUSG00000036928', 'ENSMUSG00000027419', 'ENSMUSG00000071178',\n",
    "       'ENSMUSG00000022034', 'ENSMUSG00000001403', 'ENSMUSG00000038224',\n",
    "       'ENSMUSG00000079015', 'ENSMUSG00000087594', 'ENSMUSG00000016319',\n",
    "       'ENSMUSG00000073530', 'ENSMUSG00000024552', 'ENSMUSG00000025386',\n",
    "       'ENSMUSG00000027322', 'ENSMUSG00000022285', 'ENSMUSG00000027469',\n",
    "       'ENSMUSG00000007892', 'ENSMUSG00000052854', 'ENSMUSG00000049517']\n",
    "gp_counts = Fit_GPcounts(X,Y.iloc[22:25,:])\n",
    "                         #.loc[gene_name]) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Probabilty of zeros in Fam184b and Pde1a genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aE17.5_2_09</th>\n",
       "      <th>aE17.5_2_16</th>\n",
       "      <th>aE17.5_1_11</th>\n",
       "      <th>aE17.5_3_07</th>\n",
       "      <th>aE17.5_4_06</th>\n",
       "      <th>aE17.5_3_04</th>\n",
       "      <th>aE17.5_2_11</th>\n",
       "      <th>aE17.5_1_25</th>\n",
       "      <th>aE17.5_4_01</th>\n",
       "      <th>aE17.5_4_03</th>\n",
       "      <th>...</th>\n",
       "      <th>aP18_3_12</th>\n",
       "      <th>aP60_1_11</th>\n",
       "      <th>aP60_3_05</th>\n",
       "      <th>aP15_1_15</th>\n",
       "      <th>aP60_1_13</th>\n",
       "      <th>aP60_3_08</th>\n",
       "      <th>aP60_5_16</th>\n",
       "      <th>aP18_1_17</th>\n",
       "      <th>aP60_1_10</th>\n",
       "      <th>aP60_5_05</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000167</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.996722</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.946870</td>\n",
       "      <td>0.812166</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.757297</td>\n",
       "      <td>0.635405</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000168</th>\n",
       "      <td>63.621062</td>\n",
       "      <td>68.773833</td>\n",
       "      <td>71.261578</td>\n",
       "      <td>74.481220</td>\n",
       "      <td>69.121539</td>\n",
       "      <td>49.542146</td>\n",
       "      <td>117.457081</td>\n",
       "      <td>35.679729</td>\n",
       "      <td>99.123239</td>\n",
       "      <td>14.775353</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>92.435945</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>147.767368</td>\n",
       "      <td>108.779767</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.081657</td>\n",
       "      <td>121.239603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000171</th>\n",
       "      <td>201.292870</td>\n",
       "      <td>355.829831</td>\n",
       "      <td>387.484830</td>\n",
       "      <td>273.332023</td>\n",
       "      <td>273.645543</td>\n",
       "      <td>287.506881</td>\n",
       "      <td>267.846054</td>\n",
       "      <td>159.369455</td>\n",
       "      <td>223.662693</td>\n",
       "      <td>241.849206</td>\n",
       "      <td>...</td>\n",
       "      <td>229.845437</td>\n",
       "      <td>50.631054</td>\n",
       "      <td>107.375087</td>\n",
       "      <td>85.026416</td>\n",
       "      <td>183.340994</td>\n",
       "      <td>127.295472</td>\n",
       "      <td>129.411838</td>\n",
       "      <td>164.126858</td>\n",
       "      <td>97.349098</td>\n",
       "      <td>47.857738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000184</th>\n",
       "      <td>213.808488</td>\n",
       "      <td>954.859882</td>\n",
       "      <td>555.840308</td>\n",
       "      <td>181.284478</td>\n",
       "      <td>223.461413</td>\n",
       "      <td>119.388451</td>\n",
       "      <td>217.350487</td>\n",
       "      <td>357.590170</td>\n",
       "      <td>99.123239</td>\n",
       "      <td>305.616520</td>\n",
       "      <td>...</td>\n",
       "      <td>948.654515</td>\n",
       "      <td>881.175082</td>\n",
       "      <td>1660.112216</td>\n",
       "      <td>1155.144598</td>\n",
       "      <td>1509.142659</td>\n",
       "      <td>1816.853557</td>\n",
       "      <td>1226.281526</td>\n",
       "      <td>2461.902866</td>\n",
       "      <td>799.344257</td>\n",
       "      <td>853.994745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000194</th>\n",
       "      <td>27.117174</td>\n",
       "      <td>24.918055</td>\n",
       "      <td>24.050783</td>\n",
       "      <td>87.831627</td>\n",
       "      <td>37.874816</td>\n",
       "      <td>38.983984</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>27.750900</td>\n",
       "      <td>30.499458</td>\n",
       "      <td>45.881361</td>\n",
       "      <td>...</td>\n",
       "      <td>23.851885</td>\n",
       "      <td>103.209457</td>\n",
       "      <td>174.601229</td>\n",
       "      <td>1.214663</td>\n",
       "      <td>2.736433</td>\n",
       "      <td>1.157232</td>\n",
       "      <td>1.043644</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>16.224850</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000197</th>\n",
       "      <td>4.171873</td>\n",
       "      <td>0.996722</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.837142</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.293189</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>27.104415</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>71.665122</td>\n",
       "      <td>38.310058</td>\n",
       "      <td>17.358473</td>\n",
       "      <td>82.447864</td>\n",
       "      <td>4.689339</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>144.636719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000202</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>32.891833</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>38.645916</td>\n",
       "      <td>52.077872</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>19.822071</td>\n",
       "      <td>30.499458</td>\n",
       "      <td>31.106007</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>70.104537</td>\n",
       "      <td>5.602178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>55.547115</td>\n",
       "      <td>30.265672</td>\n",
       "      <td>5.861673</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30.841653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000214</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>145.521443</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.867659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33.041080</td>\n",
       "      <td>54.435513</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000215</th>\n",
       "      <td>2.085936</td>\n",
       "      <td>926.951660</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>53.401629</td>\n",
       "      <td>11.362445</td>\n",
       "      <td>22.740657</td>\n",
       "      <td>97.697946</td>\n",
       "      <td>5125.194804</td>\n",
       "      <td>131.528913</td>\n",
       "      <td>73.876767</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28.010892</td>\n",
       "      <td>20.649273</td>\n",
       "      <td>1.368216</td>\n",
       "      <td>16.201242</td>\n",
       "      <td>1.043644</td>\n",
       "      <td>4.689339</td>\n",
       "      <td>154.676900</td>\n",
       "      <td>1157.093752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000216</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 328 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    aE17.5_2_09  aE17.5_2_16  aE17.5_1_11  aE17.5_3_07  \\\n",
       "ENSMUSG00000000167     0.000000     0.996722     0.000000     0.000000   \n",
       "ENSMUSG00000000168    63.621062    68.773833    71.261578    74.481220   \n",
       "ENSMUSG00000000171   201.292870   355.829831   387.484830   273.332023   \n",
       "ENSMUSG00000000184   213.808488   954.859882   555.840308   181.284478   \n",
       "ENSMUSG00000000194    27.117174    24.918055    24.050783    87.831627   \n",
       "ENSMUSG00000000197     4.171873     0.996722     0.000000     9.837142   \n",
       "ENSMUSG00000000202     0.000000    32.891833     0.000000    38.645916   \n",
       "ENSMUSG00000000214     0.000000   145.521443     0.000000     0.000000   \n",
       "ENSMUSG00000000215     2.085936   926.951660     0.000000    53.401629   \n",
       "ENSMUSG00000000216     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "                    aE17.5_4_06  aE17.5_3_04  aE17.5_2_11  aE17.5_1_25  \\\n",
       "ENSMUSG00000000167     0.946870     0.812166     0.000000     4.757297   \n",
       "ENSMUSG00000000168    69.121539    49.542146   117.457081    35.679729   \n",
       "ENSMUSG00000000171   273.645543   287.506881   267.846054   159.369455   \n",
       "ENSMUSG00000000184   223.461413   119.388451   217.350487   357.590170   \n",
       "ENSMUSG00000000194    37.874816    38.983984     0.000000    27.750900   \n",
       "ENSMUSG00000000197     0.000000     0.000000     3.293189     0.000000   \n",
       "ENSMUSG00000000202    52.077872     0.000000     0.000000    19.822071   \n",
       "ENSMUSG00000000214     0.000000    17.867659     0.000000     0.000000   \n",
       "ENSMUSG00000000215    11.362445    22.740657    97.697946  5125.194804   \n",
       "ENSMUSG00000000216     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "                    aE17.5_4_01  aE17.5_4_03  ...   aP18_3_12   aP60_1_11  \\\n",
       "ENSMUSG00000000167     0.635405     0.000000  ...    0.000000    0.000000   \n",
       "ENSMUSG00000000168    99.123239    14.775353  ...    0.000000    0.000000   \n",
       "ENSMUSG00000000171   223.662693   241.849206  ...  229.845437   50.631054   \n",
       "ENSMUSG00000000184    99.123239   305.616520  ...  948.654515  881.175082   \n",
       "ENSMUSG00000000194    30.499458    45.881361  ...   23.851885  103.209457   \n",
       "ENSMUSG00000000197     0.000000     0.000000  ...   27.104415    0.000000   \n",
       "ENSMUSG00000000202    30.499458    31.106007  ...    0.000000   70.104537   \n",
       "ENSMUSG00000000214    33.041080    54.435513  ...    0.000000    0.000000   \n",
       "ENSMUSG00000000215   131.528913    73.876767  ...    0.000000    0.000000   \n",
       "ENSMUSG00000000216     0.000000     0.000000  ...    0.000000    0.000000   \n",
       "\n",
       "                      aP60_3_05    aP15_1_15    aP60_1_13    aP60_3_08  \\\n",
       "ENSMUSG00000000167     0.000000     0.000000     0.000000     0.000000   \n",
       "ENSMUSG00000000168    92.435945     0.000000   147.767368   108.779767   \n",
       "ENSMUSG00000000171   107.375087    85.026416   183.340994   127.295472   \n",
       "ENSMUSG00000000184  1660.112216  1155.144598  1509.142659  1816.853557   \n",
       "ENSMUSG00000000194   174.601229     1.214663     2.736433     1.157232   \n",
       "ENSMUSG00000000197     0.000000    71.665122    38.310058    17.358473   \n",
       "ENSMUSG00000000202     5.602178     0.000000     0.000000    55.547115   \n",
       "ENSMUSG00000000214     0.000000     0.000000     0.000000     0.000000   \n",
       "ENSMUSG00000000215    28.010892    20.649273     1.368216    16.201242   \n",
       "ENSMUSG00000000216     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "                      aP60_5_16    aP18_1_17   aP60_1_10    aP60_5_05  \n",
       "ENSMUSG00000000167     0.000000     0.000000    0.000000     0.000000  \n",
       "ENSMUSG00000000168     0.000000     0.000000    1.081657   121.239603  \n",
       "ENSMUSG00000000171   129.411838   164.126858   97.349098    47.857738  \n",
       "ENSMUSG00000000184  1226.281526  2461.902866  799.344257   853.994745  \n",
       "ENSMUSG00000000194     1.043644     0.000000   16.224850     0.000000  \n",
       "ENSMUSG00000000197    82.447864     4.689339    0.000000   144.636719  \n",
       "ENSMUSG00000000202    30.265672     5.861673    0.000000    30.841653  \n",
       "ENSMUSG00000000214     0.000000     0.000000    0.000000     0.000000  \n",
       "ENSMUSG00000000215     1.043644     4.689339  154.676900  1157.093752  \n",
       "ENSMUSG00000000216     0.000000     0.000000    0.000000     0.000000  \n",
       "\n",
       "[10 rows x 328 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y.iloc[22:25,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSMUSG00000000167    0.814024\n",
       "ENSMUSG00000000168    0.274390\n",
       "ENSMUSG00000000171    0.015244\n",
       "ENSMUSG00000000184    0.000000\n",
       "ENSMUSG00000000194    0.213415\n",
       "ENSMUSG00000000197    0.368902\n",
       "ENSMUSG00000000202    0.503049\n",
       "ENSMUSG00000000214    0.896341\n",
       "ENSMUSG00000000215    0.268293\n",
       "ENSMUSG00000000216    0.969512\n",
       "dtype: float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Y.loc[gene_name][Y.loc[gene_name] == 0].count(axis=1)/len(Y.columns)\n",
    "Y.iloc[20:30,:][Y.iloc[20:30,:] == 0].count(axis=1)/len(Y.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "\n",
    "def plot():\n",
    "    plt.tick_params(labelsize='large', width=2)     \n",
    "    plt.ylabel('Gene Expression', fontsize=16)\n",
    "    plt.xlabel('Times', fontsize=16)\n",
    "    \n",
    "    if model_index == 1 or model_index == 2:\n",
    "        c = 'blue'\n",
    "\n",
    "    else:\n",
    "        c = 'salmon'\n",
    "    \n",
    "    \n",
    "    if likelihood == 'Gaussian':\n",
    "        plt.fill_between(xtest[:,0],\n",
    "                            mean[:,0] - 1*np.sqrt(var[:,0]),\n",
    "                            mean[:,0] + 1*np.sqrt(var[:,0]), alpha=0.2) # one standard deviation\n",
    "        plt.fill_between(xtest[:,0],\n",
    "                            mean[:,0] - 2*np.sqrt(var[:,0]),\n",
    "                            mean[:,0] + 2*np.sqrt(var[:,0]),color='light'+c, alpha=0.2)# two standard deviation\n",
    "    else:\n",
    "        \n",
    "        lowess = sm.nonparametric.lowess\n",
    "        \n",
    "        # one standard deviation 68%\n",
    "        percentile_16 = lowess(np.percentile(var, 16, axis=0),xtest[:,0],frac=1./5, return_sorted=False)\n",
    "        percentile_84 = lowess(np.percentile(var, 84, axis=0),xtest[:,0],frac=1./5, return_sorted=False)\n",
    "        plt.fill_between(xtest[:,0],percentile_16,percentile_84,alpha = 0.2)\n",
    "        \n",
    "        # two standard deviation 95%\n",
    "        percentile_5 = lowess(np.percentile(var, 5, axis=0),xtest[:,0],frac=1./5, return_sorted=False)\n",
    "        percentile_95 = lowess(np.percentile(var,95, axis=0),xtest[:,0],frac=1./5, return_sorted=False)\n",
    "        plt.fill_between(xtest[:,0],percentile_5,percentile_95,color='light'+c,alpha = 0.2)\n",
    "        \n",
    "    plt.plot(xtest, mean, lw=2) \n",
    "    plt.scatter(model.data[0],model.data[1], s=10, color= c, alpha=0.6) #data\n",
    "    \n",
    "    if test == 'Two_samples_test':\n",
    "        if model_index == 1  or model_index ==3:\n",
    "            plt.scatter(model.data[0][int(model.data[0].shape[0]/2)::],model.data[1][int(model.data[0].shape[0]/2)::], s=10, color= 'salmon', alpha=0.6) #data\n",
    "            plt.show()\n",
    "    \n",
    "    else:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 12%|█▎        | 1/8 [00:22<02:38, 22.63s/it]"
     ]
    }
   ],
   "source": [
    "likelihood = 'Zero_inflated_negative_binomial' \n",
    "log_likelihood = gp_counts.One_sample_test(likelihood)\n",
    "log_likelihood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#likelihood = 'Negative_binomial' \n",
    "indexes = log_likelihood.index.values.tolist() # list of genes to be plotted \n",
    "test = 'One_sample_test' # name of the test\n",
    "xtest = np.linspace(np.min(X.values),np.max(X.values),100)[:,None]\n",
    "\n",
    "params = gp_counts.load_models(indexes,test,xtest,likelihood)\n",
    "\n",
    "for i in range(len(indexes)):\n",
    "    fig = plt.figure()\n",
    "    print(indexes[i])\n",
    "    model_index = 1\n",
    "    for model,mean,var in zip(params['models'][i],params['means'][i],params['vars'][i]):\n",
    "        plot()\n",
    "        model_index = model_index + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fit GPcount with negative binomial likelihood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "likelihood = 'Negative_binomial' \n",
    "log_likelihood_nb = gp_counts.Infer_trajectory(likelihood)\n",
    "log_likelihood_nb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = gp_counts.load_models(indexes,test,xtest,likelihood)\n",
    "for i in range(len(indexes)):\n",
    "    fig = plt.figure()\n",
    "    print(indexes[i])\n",
    "    model_index = 1\n",
    "    for model,mean,var in zip(params['models'][i],params['means'][i],params['vars'][i]):\n",
    "        plot()\n",
    "        model_index = model_index + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fit GPcount with Gaussian likelihood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "likelihood = 'Gaussian'\n",
    "log_likelihood = gp_counts.Infer_trajectory(likelihood)\n",
    "log_likelihood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "params = gp_counts.load_models(indexes,test,xtest,likelihood)\n",
    "\n",
    "for i in range(len(indexes)):\n",
    "    fig = plt.figure()\n",
    "    print(indexes[i])\n",
    "    model_index = 1\n",
    "    for model,mean,var in zip(params['models'][i],params['means'][i],params['vars'][i]):\n",
    "        plot()\n",
    "        model_index = model_index + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:myenv] *",
   "language": "python",
   "name": "conda-env-myenv-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
